charms/worker/k8s/tests/unit/test_k8sd_api_manager.py

```diff
@@ -143,6 +143,53 @@ def setUp(self):
         self.mock_factory = MagicMock()
         self.api_manager = K8sdAPIManager(factory=self.mock_factory)
 
+    @patch("lib.charms.k8s.v0.k8sd_api_manager.K8sdAPIManager._send_request")
+    def test_check_k8sd_in_error(self, mock_send_request):
+        """Test bootstrap."""
+        not_found = InvalidResponseError(code=404, msg="Not Found")
+        in_error = InvalidResponseError(code=504, msg="In Error")
+        mock_send_request.side_effect = [not_found, in_error]
+
+        with self.assertRaises(InvalidResponseError) as ie:
+            self.api_manager.check_k8sd_ready()
+        mock_send_request.assert_has_calls(
+            [
+                call("/core/1.0/ready", "GET", EmptyResponse),
+                call("/cluster/1.0/ready", "GET", EmptyResponse),
+            ]
+        )
+        assert ie.exception.code == 504
+
+    @patch("lib.charms.k8s.v0.k8sd_api_manager.K8sdAPIManager._send_request")
+    def test_check_k8sd_not_found(self, mock_send_request):
+        """Test bootstrap."""
+        not_found = InvalidResponseError(code=404, msg="Not Found")
+        mock_send_request.side_effect = [not_found, not_found]
+
+        with self.assertRaises(K8sdConnectionError):
+            self.api_manager.check_k8sd_ready()
+        mock_send_request.assert_has_calls(
+            [
+                call("/core/1.0/ready", "GET", EmptyResponse),
+                call("/cluster/1.0/ready", "GET", EmptyResponse),
+            ]
+        )
+
+    @patch("lib.charms.k8s.v0.k8sd_api_manager.K8sdAPIManager._send_request")
+    def test_check_k8sd_ready(self, mock_send_request):
+        """Test bootstrap."""
+        not_found = InvalidResponseError(code=404, msg="Not Found")
+        success = EmptyResponse(status_code=200, type="test", error_code=0)
+        mock_send_request.side_effect = [not_found, success]
+
+        self.api_manager.check_k8sd_ready()
+        mock_send_request.assert_has_calls(
+            [
+                call("/core/1.0/ready", "GET", EmptyResponse),
```

Comment:
Shouldn't this be InvalidResponseError? Because of `mock_send_request.side_effect = [not_found, success]` I would expect the first one to be not found, but I may be misunderstanding.

Reply:
In this test case the first endpoint isn't found, but the second is.  Because the second succeeds -- there's no exception right?

Reply:
EmptyResponse is defined as having a 200 status code. If the first endpoint isn't found, why are we expecting a 200 status code?

Reply:
Ahh... no 

`EmptyResponse` is the expected "happy" return type.  We'd like to send a request, and get an `EmptyResponse` back. That "third" argument is the expected return type.  

So, yeah `self._send_request(...)` wants to see if the cluster is ready, and see an `EmptyResponse` instance come back.

If we get a status code in the 2xx range, we'll parse the response, otherwise we get an `InvalidResponseError`